src/mongoNotes.js

```diff
@@ -0,0 +1,306 @@
+   * {
+   *   blocksInteraction: false,
+   *   componentStyle: "banner",
+   *   context: ["artikkelit", "muuntaja"],
+   *   endDate: new Date("2024-12-30T12:30:15.002"),
+   *   isDismissible: true,
+   *   messageStyle: "info",
+   *   messageText: "Test server notification message",
+   *  }
+   * @returns {void}
+   */
+  async function addNoteItem(noteItem) {
+    logger.info(`MongoNotes: Adding one note item ${JSON.stringify(noteItem)}`);
+
+    if (isNotObject(noteItem)) {
+      logger.debug('MongoNotes: NoteItem parameter is not object');
+      throw new ApiError(httpStatus.INTERNAL_SERVER_ERROR, 'NoteItem data is not valid');
+    }
+
+    const newNoteItem = {
+      blocksInteraction: validate(noteItem.blocksInteraction, isBoolean),
+      componentStyle: validate(noteItem.componentStyle, isValidComponentStyle),
+      context: validate(noteItem.context, isValidContext),
+      endDate: validate(noteItem.endDate, isValidEndDate),
+      isDismissible: validate(noteItem.isDismissible, isBoolean),
+      messageStyle: validate(noteItem.messageStyle, isValidMessageStyle),
+      messageText: validate(noteItem.messageText, isValidMessageText)
+    };
+
+    if (hasUndefinedProperty(newNoteItem)) {
+      logger.debug('MongoNotes: NoteItem data did not pass validation');
+      throw new ApiError(httpStatus.INTERNAL_SERVER_ERROR, 'NoteItem data is not valid');
+    }
+
+    // '_id' property is automatically generated by mongo node driver
+    // and added to noteItem before insert to collection.
+    // type of '_id' is ObjectId.
+    const result = await db.collection(collection).insertOne(newNoteItem);
+
+    if (result.acknowledged) {
+      logger.info(`New ${noteItem.componentStyle} note item added with ${noteItem.messageStyle} message: ${noteItem.messageText}`);
+
+      const {_id: _, ...rest} = newNoteItem;
+      return {id: result.insertedId.toString(), ...rest};
+    }
+
+    throw new ApiError(httpStatus.INTERNAL_SERVER_ERROR, `MongoNotes had error while adding a note ${noteItem}`);
+
+
+    /**
+     * Validate property with validator function
+     * @param {*} property
+     * @param {function} validator
+     * @returns {(*|undefined)}
+     */
+    function validate(property, validator) {
+      if (property === undefined) {
+        return undefined;
+      }
+      return validator(property) ? property : undefined;
+    }
+
+    /**
+     * Checks if noteItem is Object
+     * @param {*} noteItem
+     * @returns {Boolean}
+     */
+    function isNotObject(noteItem) {
+      return typeof noteItem !== 'object' || Object.keys(noteItem).length === 0 || Object.getPrototypeOf(noteItem) !== Object.prototype;
+    }
+
+    /**
+     * Validates if property is Boolean
+     * @param {*} property
+     * @returns {Boolean}
+     */
+    function isBoolean(property) {
+      return typeof property === 'boolean';
+    }
+
+    /**
+    * Validates if every app in list is valid context
+    * @param {*} appsList
```

Comment:
Incoming param should be array of strings? `[String]`

Reply:
Param comments updated